src/rust/engine/src/rule_graph.rs

```diff
@@ -454,6 +463,99 @@ pub struct RuleGraph {
+  let rpr_val = externs.project_ignoring_type(&value, "__repr__");
+
+  let invoke_result  = externs.invoke_runnable(&Runnable{func: rpr_val, args:vec![], cacheable:false})
+                              .expect("string from calling repr");
+  externs.val_to_str(&invoke_result)
+}
+
+
+fn function_str(externs: &Externs, func: &Function) -> String {
+  let as_val = to_val_from_func(externs, func);
+  val_name(externs, &as_val)
+}
+
+
+fn type_str(externs: &Externs, type_id: TypeId) -> String {
+  if type_id == ANY_TYPE {
+    "Any".to_string()
+  } else {
+    let as_val = to_val_from_id(externs, type_id.0);
+    val_name(externs, &as_val)
+  }
+}
+
+fn val_name(externs: &Externs, val: &Value) -> String {
+  externs.project_str(val, "__name__")
+}
+
+fn selector_str(externs: &Externs, selector: &Selector) -> String {
+  match selector {
+    &Selector::Select(ref s) => format!("Select({})", type_constraint_str(externs, s.product)).to_string(), // TODO variant key
+    &Selector::SelectDependencies(ref s) => format!("SelectDependencies({}, {}, {}field_types=({},){})",
+                                                    type_constraint_str(externs, s.product),
+                                                    type_constraint_str(externs, s.dep_product),
+                                                    if s.field == "dependencies" { "".to_string() } else {format!("{:?}, ", s.field)},
+                                                    s.field_types.iter()
+                                                                 .map(|&f| type_str(externs, f))
+                                                                 .collect::<Vec<String>>()
+                                                                 .join(", "),
+                                                    if s.transitive { ", transitive=True" } else { "" }.to_string()
+
+    ),
+    &Selector::SelectProjection(ref s) => format!("SelectProjection({}, {}, ({:?},), {})",
+                                                  type_constraint_str(externs, s.product),
+                                                  type_str(externs, s.projected_subject),
+                                                  s.field,
+                                                  type_constraint_str(externs, s.input_product),
+    ),
+    &Selector::SelectLiteral(ref s) => format!("SelectLiteral({:?})", s),
+    &Selector::Task(ref s) => format!("{:?}", s)
+  }
+}
+
+fn entry_str(externs: &Externs, entry: &Entry) -> String {
+  match entry {
+    &Entry::InnerEntry(ref inner) => {
+      format!("{} of {}", task_display(externs, &inner.rule), type_str(externs, inner.subject_type))
+    }
+    other => {
+      format!("{:?}", other)
+    }
+  }
+}
+
+fn task_display(externs: &Externs, task: &Task) -> String {
+  let product = type_constraint_str(externs, task.product);
+  let mut clause_portion = task.clause.iter().map(|c| selector_str(externs, c)).collect::<Vec<_>>().join(", ");
```

Comment:
hm. I remember it complaining. join is defined for slices, an iterator isn't a slice. Also, I think it complains without `::<>` because the FromIterator can't be inferred.

https://doc.rust-lang.org/std/slice/trait.SliceConcatExt.html